services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,572 @@
+
+
+import com.futurewei.alcor.common.entity.*;
+import com.futurewei.alcor.common.rest.IpAddressRest;
+import com.futurewei.alcor.common.rest.MacAddressRest;
+import com.futurewei.alcor.common.rest.VpcRest;
+import com.futurewei.alcor.portmanager.exception.*;
+import com.futurewei.alcor.portmanager.repo.PortRepository;
+import com.futurewei.alcor.portmanager.rollback.*;
+import com.futurewei.alcor.portmanager.service.PortService;
+import com.futurewei.alcor.portmanager.utils.Ipv4AddrUtil;
+import com.futurewei.alcor.portmanager.utils.Ipv6AddrUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+@ComponentScan(value="com.futurewei.alcor.common.rest")
+public class PortServiceImpl implements PortService {
+    private static final Logger LOG = LoggerFactory.getLogger(PortServiceImpl.class);
+
+    @Autowired
+    private PortRepository portRepository;
+
+    @Autowired
+    private IpAddressRest ipAddressRest;
+
+    @Autowired
+    private MacAddressRest macAddressRest;
+
+    @Autowired
+    private VpcRest vpcRest;
+
+    private int getIpVersion(String ipAddress) throws Exception {
+        if (Ipv4AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV4.getVersion();
+        } else if (Ipv6AddrUtil.formatCheck(ipAddress)) {
+            return IpVersion.IPV6.getVersion();
+        } else {
+            throw new IpAddrInvalidException();
+        }
+    }
+
+    private String getRangeIdBySubnetId(String subnetId, int ipVersion) throws Exception {
+        if (IpVersion.IPV4.getVersion() == ipVersion) {
+            return subnetId; //FIXME:return the right rangeId get from subnet manager
+        } else if (IpVersion.IPV6.getVersion() == ipVersion) {
+            return subnetId; //FIXME:return the right rangeId get from subnet manager
+        }
+
+        throw new IpVersionInvalidException();
+    }
+
+    private void verifyIpAddresses(List<PortState.FixedIp> fixedIps, Stack<PortStateRollback> rollbacks) throws Exception {
+        for (PortState.FixedIp fixedIp: fixedIps) {
+            int ipVersion = getIpVersion(fixedIp.getIpAddress());
+            String rangeId = getRangeIdBySubnetId(fixedIp.getSubnetId(), ipVersion);
+
+            IpAddrRequest result = ipAddressRest.allocateIpAddress(rangeId, fixedIp.getIpAddress());
+
+            AllocateIpAddrRollback ipAddressRollback = new AllocateIpAddrRollback(ipAddressRest);
+            ipAddressRollback.putAllocatedIpAddress(result);
+            rollbacks.push(ipAddressRollback);
+        }
+    }
+
+    private String getRangeIdForPort(String vpcId) {
+        return "range1";
+    }
+
+    private void allocateIpAddress(PortState portState, Stack<PortStateRollback> rollbacks) throws Exception {
```

Comment:
The idea of using a rollback stack to keep track of potential rollback call is good. 

The issue, however, is that it might not work in current implementation. If an exception is thrown before rollbacks.push(ipAddressRollback), then the stack doesn't have ipAddressRollback therefore won't be able to trigger rollback during exception handling.

Reply:
Yes, we can consider it in two cases. First, if an exception occurs before receiving the rest response, there is no need to roll back. Second, if an exception occurs after receiving the rest response, it needs to be rolled back, so we need to ensure that there is no exception between the rest response and rollbacks.push (). In fact, rollbacks.push () is executed after receiving the rest response, so we can assume that there will be no exception between them.

Reply:
I want to deal with the rollbacks of various operations in a unified way by a common method. This method does not need to judge whether the operation needs to be rolled back, and does not need to care about what to be rolled back, this should be concerned by the operation itself. There may be a better way, do you have any good suggestions?

Reply:
@chenpiaoping I miss your previous comment in this thread. Unifying the rollback is a good idea but depending on how we are going to do it.

Regarding the first case, let us say Service A => Service B, and the modify (POST or PUT or DELETE) request has been sent but no yet received a response, an exception is received at this time. It is still necessary that Service A send roolback to Service B.

Reply:
Oh, this is indeed a big problem. Does subnet manager have the same problem?

Reply:
Submitted a commit for this problem: https://github.com/futurewei-cloud/alcor/pull/180/commits/0a4c96a05592534af52680abe72f9d5d6f364d28

Reply:
https://github.com/futurewei-cloud/alcor/pull/180/commits/0a4c96a05592534af52680abe72f9d5d6f364d28，This commit is going to solve this problem.

Reply:
Thanks let me check it.